doc/L3_performance_and_scaling_enchancements_HLD.md

```diff
@@ -0,0 +1,228 @@
+
+### Table 2: IPv4 prefix route programming time
+
+| Routes                  | time taken on AS7712(Tomahawk) |
+| ----------------------- | ------------------------------ |
+| 10k IPv4 prefix routes  | 11 seconds                     |
+| 30k IPv4 prefix routes  | 30 seconds                     |
+| 60k IPv4 prefix routes | 48 seconds                     |
+| 90k IPv4 prefix routes  | 68 seconds                     |
+
+
+
+### Table 3: IPv6 prefix route programming time
+
+| Routes                             | time taken on AS7712(Tomahawk) |
+| ---------------------------------- | -----------------------------  |
+| 10k IPv6 route with prefixes > 64b | 11 seconds                     |
+| 30k IPv6 route with prefixes >64b  | 30 seconds                     |
+
+
+#### 2.2.1.1 Proposed optimizations for reducing the route programming time.
+
+- <u>Using sairedis bulk route APIs</u>
+
+  In SONiC architecture, routeorch in Orchagent processes route table updates from the APP_DB and calls the sairedis APIs to put these routes in ASIC_DB. 
+  Currently Orchagent processes each route and puts in the ASIC_DB one at a time. The Redis pipelining allows for some level of bulking when putting entries in ASIC_DB but still one DB message is generated for every route.
+
+  Further bulking can be done by using the sairedis bulk APIs for route creation and deletion.
+
+  By using Sairedis bulk APIs, orchagent will call these APIs with a list of routes and their attributes.
+  The meta_sai layer in sairedis iterates over this route list and creates the meta objects for every route but only one Redis DB message will be generated for the route list. 
+  Therefore using the sairedis bulk APIs reduces the number of Redis messages. 
+  
+  The ASIC doesn't support bulk route creation/deletion,so syncd still processes one route at time and updates the ASIC.
+
+  So, the saving achieved by using bulk APIs will be number of Redis message generated.
+
+  Bulking of Route updates will be enabled in Orchagent. Orchagent will bulk 64 updates and send to Sairedis. 
+  A new timer will be introduced in orchagent to flush the outstanding updates every second.
+
+- <u>Optimization in Fpmsyncd</u>
+
+  Fpmsyncd listens to Netlink messages for Route add/delete messages and updates the APP_DB.
+  Current behaviour in Fpmsyncd
+   - When fpmsyncd get a route first it tries to get the master device name from the rt_table attribute of route object.
+     This is done to check if the route belongs to VNET_ROUTE_TABLE.
+  - To get a Master device name, fpmsyncd does a lookup in its local link cache. 
+  - If the lookup in the local link cache fails, fpmsyncd updates cache by getting the configured links from the kernel.
+  
+  The problem here is if there are no VNET present on the system, this lookup will always fail and cache is updated for every route.
+  This seems to slow down the rate which route programed for the global route table. 
```

Comment:
@prsunny , can you take a look at this problem?

Reply:
This is fixed as part of the PR https://github.com/Azure/sonic-swss/pull/1048